# Data Preprocessing Before Class


In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, MinMaxScaler, StandardScaler
from pickle import dump

In [ ]:
!gdown --id 1iiZ3i1x_XKvuiD4Ff_qcwXbrQNcHVSR6

Downloading...
From: https://drive.google.com/uc?id=1iiZ3i1x_XKvuiD4Ff_qcwXbrQNcHVSR6
To: /content/filtered_data.csv
100% 63.9M/63.9M [00:00<00:00, 104MB/s] 


In [ ]:
data = pd.read_csv("filtered_data.csv")

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [ ]:
data.head()

city state house         street  ... fireplace water  sewer  housePrice
0  Lilburn    GA   510  Village Green  ...   Masonry   Yes    NaN      200300
1  Lilburn    GA   540  Village Green  ...   Masonry   Yes    NaN      193800
2  Lilburn    GA  5133          Arbor  ...   Masonry   Yes    NaN      197500
3  Lilburn    GA   215          Shunn  ...   Masonry   Yes    NaN      204700
4  Lilburn    GA   195          Shunn  ...   Masonry   Yes    NaN      195900

[5 rows x 29 columns]

In [ ]:
# Dropping "state", "zip4", "unit", "foundation", "sewer", "water", "fireplace", "heating", "airConditioning", "architecturalStyle"
data.drop(columns=["state","zip4", "unit", "foundation", "sewer", "water", "fireplace", "heating", "airConditioning", "architecturalStyle"], inplace=True)

In [ ]:
data.head()

city house         street  ... bedrooms  baths  housePrice
0  Lilburn   510  Village Green  ...      4.0    2.0      200300
1  Lilburn   540  Village Green  ...      3.0    2.0      193800
2  Lilburn  5133          Arbor  ...      3.0    2.0      197500
3  Lilburn   215          Shunn  ...      4.0    3.0      204700
4  Lilburn   195          Shunn  ...      3.0    2.0      195900

[5 rows x 19 columns]

In [ ]:
# Counting the number of NaN in the dataframe
data.isna().sum()

city                    773
house                 10984
street                10984
streetSuffix          12070
zip                    1332
latitude                  0
longitude                 0
landUseDescription        0
zoningDescription     11795
lotSizeAcres            154
lotSizeSquareFeet       154
lotTopography         41216
condition             17444
yearBuilt             17320
totalStories          19504
totalRooms            29911
bedrooms              29795
baths                 30039
housePrice                0
dtype: int64

In [ ]:
# Now we remove all the NaNs belonging to categorical features, since we have huge amount of data
cols = list(data.columns)
data = data.dropna(subset=cols)

In [ ]:
# Checking if there is any NaN or not
data.isna().sum()

city                  0
house                 0
street                0
streetSuffix          0
zip                   0
latitude              0
longitude             0
landUseDescription    0
zoningDescription     0
lotSizeAcres          0
lotSizeSquareFeet     0
lotTopography         0
condition             0
yearBuilt             0
totalStories          0
totalRooms            0
bedrooms              0
baths                 0
housePrice            0
dtype: int64

In [ ]:
data.info()

In [ ]:
data.describe()

zip       latitude  ...          baths    housePrice
count  217026.000000  217026.000000  ...  217026.000000  2.170260e+05
mean    30109.287049      33.952778  ...       2.767309  2.428842e+05
std       152.930324       0.085889  ...       0.840779  1.302617e+05
min     30011.000000      33.755535  ...       1.000000  0.000000e+00
25%     30043.000000      33.890090  ...       2.000000  1.732000e+05
50%     30046.000000      33.946106  ...       3.000000  2.153000e+05
75%     30092.000000      34.013120  ...       3.000000  2.780000e+05
max     30620.000000      34.167661  ...      13.000000  1.100430e+07

[8 rows x 11 columns]

In [ ]:
data.head()

city house         street  ... bedrooms  baths  housePrice
0  Lilburn   510  Village Green  ...      4.0    2.0      200300
2  Lilburn  5133          Arbor  ...      3.0    2.0      197500
3  Lilburn   215          Shunn  ...      4.0    3.0      204700
4  Lilburn   195          Shunn  ...      3.0    2.0      195900
5  Lilburn   315          Shunn  ...      3.0    3.0       96000

[5 rows x 19 columns]

In [ ]:
def label_encode(data, column):
  data_copy = data.copy()
  data_copy[column] = data_copy[column].str.lower()
  le = LabelEncoder()
  le.fit(data_copy[column])
  data_copy[column] = le.fit_transform(data_copy[column])
  
  noClasses = le.classes_
  encodings = list(range(len(noClasses)))
  encoding_map = dict(zip(noClasses, encodings))

  return data_copy, encoding_map

In [ ]:
categorical_cols = ["city", "street", "streetSuffix", "landUseDescription", "zoningDescription", "lotTopography", "condition"]
numerical_cols =	["house", "zip", "latitude", "longitude", "lotSizeAcres",	"lotSizeSquareFeet",	"yearBuilt",	"totalStories",	"totalRooms",	"bedrooms",	"baths"]

In [ ]:
encodings_database = dict()
for column in categorical_cols:
  data, encoding_map = label_encode(data, column)
  encodings_database[column] = encoding_map

In [ ]:
encodings_database.keys()

dict_keys(['city', 'street', 'streetSuffix', 'landUseDescription', 'zoningDescription', 'lotTopography', 'condition'])

In [ ]:
encodings_database.values()

dict_values([{'atlanta': 0, 'auburn': 1, 'bethlehem': 2, 'braselton': 3, 'buford': 4, 'conyers': 5, 'dacula': 6, 'duluth': 7, 'grayson': 8, 'hoschton': 9, 'lawrenceville': 10, 'lilburn': 11, 'lithonia': 12, 'loganville': 13, 'norcross': 14, 'peachtree corners': 15, 'snellville': 16, 'stone mountain': 17, 'suwanee': 18, 'tucker': 19}, {'1st': 0, '2nd': 1, '3rd': 2, '5th': 3, 'aaron': 4, 'abba': 5, 'abbey': 6, 'abbey brook': 7, 'abbey hill': 8, 'abbey park': 9, 'abbie kilgore': 10, 'abbotts bridge': 11, 'abbotts oak': 12, 'abbotts pointe': 13, 'abby': 14, 'abby grace': 15, 'abercorn': 16, 'aberdeen': 17, 'abernathy': 18, 'aberrone': 19, 'abigail': 20, 'abilene': 21, 'abinger': 22, 'abington': 23, 'abri': 24, 'academic': 25, 'academy': 26, 'accord': 27, 'ace mcmillian': 28, 'acorn oak': 29, 'acuba': 30, 'adah': 31, 'adair': 32, 'adair hill': 33, 'adams': 34, 'adams brook': 35, 'adams farm': 36, 'adams lake': 37, 'adams landing': 38, 'adams mill': 39, 'adams pointe': 40, 'add': 41, 'addies

In [ ]:
for key, values in encodings_database.items():
  print(f"{key}: {len(values)}")

city: 20
street: 8153
streetSuffix: 63
landUseDescription: 16
zoningDescription: 85
lotTopography: 5
condition: 5


In [ ]:
assert len(encodings_database.keys()) == len(encodings_database.values())

In [ ]:
data.head()

city house  street  streetSuffix  ...  totalRooms  bedrooms  baths  housePrice
0    11   510    7536            13  ...         6.0       4.0    2.0      200300
2    11  5133     236            29  ...         7.0       3.0    2.0      197500
3    11   215    6476            61  ...         7.0       4.0    3.0      204700
4    11   195    6476            61  ...         6.0       3.0    2.0      195900
5    11   315    6476            61  ...         8.0       3.0    3.0       96000

[5 rows x 19 columns]

In [ ]:
# NORMALIZATION:
# Now we will be using MinMax Scalar for the Categorical Features
categorical_data = data[categorical_cols]

minMaxScalar = MinMaxScaler()
data[categorical_cols] = minMaxScalar.fit_transform(categorical_data)

In [ ]:
data.head()

city house    street  ...  bedrooms  baths  housePrice
0  0.578947   510  0.924436  ...       4.0    2.0      200300
2  0.578947  5133  0.028950  ...       3.0    2.0      197500
3  0.578947   215  0.794406  ...       4.0    3.0      204700
4  0.578947   195  0.794406  ...       3.0    2.0      195900
5  0.578947   315  0.794406  ...       3.0    3.0       96000

[5 rows x 19 columns]

In [ ]:
# We use Standard Scalar to scale our numerical data since we have the continuous values without any maximum and minimum values.
numerical_data = data[numerical_cols]

standardScalar = StandardScaler()
data[numerical_cols] = standardScalar.fit_transform(numerical_data)

In [ ]:
data.head()

city     house    street  ...  bedrooms     baths  housePrice
0  0.578947 -1.175826  0.924436  ...  0.688317 -0.912619      200300
2  0.578947  1.595196  0.028950  ... -0.637291 -0.912619      197500
3  0.578947 -1.352649  0.794406  ...  0.688317  0.276757      204700
4  0.578947 -1.364637  0.794406  ... -0.637291 -0.912619      195900
5  0.578947 -1.292709  0.794406  ... -0.637291  0.276757       96000

[5 rows x 19 columns]

In [ ]:
data.to_csv("Processed Data.csv", index=False)

In [ ]:
dump(minMaxScalar, open("minMaxScalar.pkl", "wb"))
dump(standardScalar, open("standardScalar.pkl", "wb"))

#NOTES

**Removing Columns:**

* Removed columns **unit**, **foundation**, **sewer**, **architecturalStyle** because of huge amount of NaN values:
  * *unit*: 260174
  * *foundation*: 239294
  * *sewer*: 115998
  * *architecturalStyle*: 83644
* Removed **zip4**: Since it was found less correlation with the housing price: -0.009316
* Removed **state**: Since we have the data of only the state GA, out model won't get affected with the state. We can add it only if we have the data from multiple states.
* Removed **water**: Since it has only given "Yes" and "NaN" which is unclear.
* Removed **Fireplace**: Since it has only single value "Masonry"
* Removed **Heating**: Since 99% data is labelled as "Forced Air"
* Removed **airConditioning**: Since 99% data is labelled as "Yes"

**Dropped NaN for the following features features**:

The features are:
  * **Categorical Features**:
    1.  *City*
    2.  *Street Name*
    3.  *Street Suffix*
    4.  *Land Use Description*
    5.  *Zoning Description*
    6.  *Lot Topography*
    7.  *Condition*

  * **Numerical Features**
    1.  *Zip*
    2.  *Latitude*
    3.  *Longitude*	
    4.  *LotSizeAcres*
    5.  *LotSizeSquareFeet*
    6.  *Year Built*
    7.  *Total Stories*
    8.  *Total Rooms*
    9.  *Bedrooms*
    10. *Baths*
    11. *House Number*
  
  * **Target**
    1. *housePrice*

**For our regression model, we have followings:**
  * *Categorical Features*: 7
  * *Numerical Features*: 11
  * *Target*: 1
  * *Total Data*: 217026


Now we encode the categorical features into numbers as follows:

**Encoding Categorical Features:**

* Converted the data into lower case
* Used **Label Encoding** to assign the number for the street as provided in our dataset (starting from 0)
* Thus, as per our data, the categorical features are in the following inclusive ranges:
  * *City*: [0 - 19]
  * *Street*: [0 - 8152]
  * *Street Suffix*: [0 - 62]
  * *Land Use Description*: [0 - 15]
  * *Zoning escription*: [0 - 84]
  * *Lot Topography*: [0 - 4]
  * *Condition*: [0 - 4]

**Normalization**

* For the *categorical features*, we have the minimum and maximum values since we have indexed from (0 - maximum). Therefore, we use the **MinMaxScaling** to scale the categorical data between 0 - 1.

* For the *numerical features*, the values are continuous and we do not have the maximum and minimum values. Therefore, we use the Standard Scaling, which computes the mean and variance based on the standard scaling technique.

* Finally, we dump the scaling models as the pickle files which can be used to scale the values of the features on predictions.